www/js/survey/enketo-survey-service.js

```diff
@@ -0,0 +1,213 @@
+  const DATA_KEY = 'manual/survey_response';
+  const ENKETO_SURVEY_CONFIG_PATH = 'json/enketoSurveyConfig.json';
+  /** @type {EnketoSurveyState} _state */
+  let _state;
+
+  /**
+   * _lazyLoadConfig load enketo survey config. If already loaded, return the cached config
+   * @returns {Promise<EnketoSurveyConfig>} enketo survey config
+   */
+  function _lazyLoadConfig() {
+    if (_state.config !== null) {
+      return Promise.resolve(_state.config);
+    }
+    return $http.get(ENKETO_SURVEY_CONFIG_PATH).then(configRes => {
+      _state.config = configRes.data;
+      return _state.config;
+    });
+  }
+
+  /**
+   * _loadForm load enketo form
+   * @param {{
+   *  instanceStr?: string;
+   *  submitted?: boolean;
+   *  external?: string[];
+   *  session?: object;
+   * }} opts enketo form options
+   * @returns {string[]} load errors
+   */
+  function _loadForm(opts = {}) {
+    const formSelector = 'form.or:eq(0)';
+    const data = {
+      // required string of the default instance defined in the XForm
+      modelStr: _state.loaded.model,
+      // optional string of an existing instance to be edited
+      instanceStr: opts.instanceStr || null,
+      submitted: opts.submitted || false,
+      external: opts.external || [],
+      session: opts.session || {}
+    };
+    _state.form = new $window.FormModule(formSelector, data, {});
+    return _state.form.init();
+  }
+
+  /**
+   * _resetState reset state object to its initial value
+   */
+  function _resetState() {
+    _state = {
+      config: null,
+      name: null,
+      form: null,
+      formLocation: null,
+      loaded: {
+        form: null,
+        model: null,
+      },
+      opts: {
+        trip: null,
+        uuid: null,
+      },
+    };
+  }
+
+  /**
+   * _restoreAnswer restore the most recent answer for the survey
+   * @param {EnketoAnswer[]} answers survey answers
+   * @returns {string} answer string
+   */
+  function _restoreAnswer(answers) {
+    let answer = null;
+    answers = EnketoSurveyAnswer.filterByNameAndVersion(_state.name, answers);
+    if (!answers.length) {
+      return null;
+    }
+    if (_state.opts.trip) {
+      answer = DiaryHelper.getUserInputForTrip(_state.opts.trip, undefined, answers);
+      if (answer) {
+        return answer.data.xmlResponse;
+      }
+    }
+    if (_state.opts.uuid) {
+      answers = answers.filter(answer => answer.data.user_uuid === _state.opts.uuid);
+      answer = answers.length ? answers[0] : null;
+    }
+    return answer ? answer.data.xmlResponse : null;
+  }
+
+  /**
+   * _saveData save survey answer data
+   * @returns {Promise<EnketoAnswerData>} answer data
+   */
+  function _saveData() {
+    /** @type {DOMParser} xmlParser */
+    const xmlParser = new $window.DOMParser();
+    const xmlResponse = _state.form.getDataStr();
+    const xmlDoc = xmlParser.parseFromString(xmlResponse, 'text/xml');
+
+    const data = {
+      label: EnketoSurveyAnswer.resolveLabel(_state.name, xmlDoc),
```

Comment:
I see this reference to "Note that the label is resolved at the answer time."

Reply:
I changed the flow to make it easier to get the answer when we call `populateInputFromTimeline. Thus avoid further processing.

Reply:
again, this is a note to self 😄